### ISSUE TITLE
Calculation of "Tage seit letzter Risiko-Begegnung" is wrong

### ISSUE THREAD

**Issue by vschoech** [corona-warn-app/cwa-app-ios#1032]
## Avoid duplicates

* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is not already reported in another issue

## Describe the bug

"Tage seit letzter Risiko-Begegnung" are calculated differently at different times of the same day. It seems that sometimes, in the morning, it calculates 1 less than in the evening. I am 99% sure that the higher number, which is shown later in the day, is the correct one.

Today I took two screenshots to illustrate the issue. I noticed, that there was an update of "infected data" in between the screenshots. Maybe the "morning vs. evening" is actually "before vs. after the daily update"?

## Expected behaviour

According to expectation based on communication between humans, the displayed number must not change within one day. Accuracy/unambiguousness is very much desirable, because indicating the wrong day can raise a lot of questions and point into wrong directions (e.g., if one day I was in the office, and the other day I was strictly at home).

## Steps to reproduce the issue

First of all, you need to register a "Risiko-Begegnung". Then, repeatedly look at the app display over the course of a day. The following screenshots were taken today, but I noticed the problem already a few times before:

![2020-08-10 07-15-58](https://user-images.githubusercontent.com/44493453/89767320-fe840c00-daf9-11ea-9766-b8999bf98de4.jpeg)

![2020-08-10 10-48-56](https://user-images.githubusercontent.com/44493453/89767316-fdeb7580-daf9-11ea-9549-f17b4cd53ebb.jpeg)

## Technical details

- iOS Version: 13.6
- Device: iPhone Xr

## Additional context

Reported the problem here already:
https://pandemie-forum.de/t/corona-warn-app-technische-probleme-und-updates/600/43?u=volker.berlin

**Comment by Ein-Tim** [comment]
Maybe a very stupid question, but couldn't it be that you had the last encounter at f.e. 8 AM, so before 8 AM only 11 days are shown and after 8 Am 12?
Just a guess without looking deeper into it.

**Comment by vschoech** [comment]
Without looking into the code, it could be anything.

But even if the bug is as you describe, it is still a bug. If the message says days, it should be days that are counted. If the app counts hours or milliseconds, the message should say hours or milliseconds. It is not in compliance with general language understanding to lump any consecutive 24 hours together and call it a "day".

I'm trying to make the point that the current behavior is bound to be confusing to normal people (outside the software development bubble). This is irrespective of why it is as it is.

**Comment by Ein-Tim** [comment]
You're right, also imho the days shouldn't change within one day, just wanted to give a guess why this happens. So or so we need a developer who takes a look at this

**Comment by Marco2907** [comment]
Hello @vschoech thanks for sharing your input. I will keep this issue to our development community and will give you an update as soon as possible. 

Thank you.

Best,
MP

Corona-Warn-App Open Source Team

---
Internal Tracking ID: EXPOSUREAPP-2406

**Comment by ir-fuel** [comment]
From what I understand there is no real "calculation" going on.

What happens is the following:
once approximately every 24h the exposure check is done. This downloads the new keys and does the internal matching (in the iOS API) of keys with the keys you exchanged. One of the results of this process (which is completely black boxed) is `daysSinceLastExposure`.

This is the value you see on screen. This value is returned by Apple and in no way related to the time on your device. 

Let's say the exposure check runs yesterday at 5am. It detects an exposure 4 days ago. It will show "4 days ago". Until the exposure check runs again it will keep on saying "4 days ago". If the exposure check today runs at 6am (remember, it's Apple that schedules this, so you'll never have it running at exactly every 24h), until 6am it will keep on saying "4 days ago". After the check runs again the value returned by the API will be "5 days ago".

This is the technical reason why it's shown like this.

The obvious (and easy?) solution to this is to just do -1 to the shown "days ago" if the last exposure check ran before midnight.

**Comment by vschoech** [comment]
Or +1, actually? ;)

**Comment by ir-fuel** [comment]
No idea what you mean 😇

**Comment by daimpi** [comment]
This issue will probably also be somewhat mitigated once we get CWA performing [multiple checks per day](https://github.com/corona-warn-app/cwa-backlog/issues/2) 🙂.

**Comment by ir-fuel** [comment]
Here is the (untested) solution:

in `ExposureDetectionViewController+DynamicTableViewModel.swift`

```
	static func riskLastExposure(text: String, image: UIImage?) -> DynamicCell {
		.risk { viewController, cell, _ in
		
                       // add offsets if not calculated today
			var daysSinceLastExposure = viewController.state.risk?.details.daysSinceLastExposure ?? 0
			
			if let date: Date = viewController.state.risk?.details.exposureDetectionDate {
				let calendar = Calendar.current
				let exposureDetectionDate = calendar.startOfDay(for: date)
				let today = calendar.startOfDay(for: Date())
				let components = calendar.dateComponents([.day], from: exposureDetectionDate, to: today)
				
				if let days = components.day {
					daysSinceLastExposure += days
				}
			}
			
			cell.textLabel?.text = .localizedStringWithFormat(text, daysSinceLastExposure)
			cell.imageView?.image = image
		}
	}
```

**Comment by ir-fuel** [comment]
@vschoech also be aware that this value does not represent what you might think it represents.

One might think this is the moment you had your latest high risk exposure, but this is not the case:


```
The values for this property represent the number of days since the most recent exposure, going backwards. For example, a value of 0 means that the person was exposed today, 1 was yesterday, 2 was two days ago, and so on.
This property is only valid if matchedKeyCount > 0.
The framework computes this property's value across all matching exposures, not just exposures that meet or exceed minimumRiskScore or minimumRiskScoreFullRange.
```

What this means is, this is the day when you had any exposure, being it low or high risk. So it could state '2 days ago' but the real high risk exposure you had was 5 days ago.

**Comment by vschoech** [comment]
@ir-fuel Thank you for the additional information! I think that for most practical purposes the actual meaning is close enough to the meaning that people expect when they see the value. Anyway, for those who wonder, it would be nice to have this detailed explanation available in the app, e.g., as part of the page that opens when I click/tap on the green/red indicator box.

**Comment by daimpi** [comment]
Anke domscheit-berg is also [irritated](https://twitter.com/anked/status/1309455001455153152) by this behavior and like OP [suggests](https://twitter.com/anked/status/1309433648593207297) using an absolute date for the time of exposure.

**Comment by RubenKelevra** [comment]
It's quite bad to not explicitly stating the time frame when the contacts occurred, since this way you could narrow it down how the contact occurred.

Like if you took the bus but it was empty but the app recorded an encounter on a traffic light stop to a passenger in a different vehicle might be a completely different risk case than a meeting in an office at work.

**Comment by daimpi** [comment]
@RubenKelevra the exact timestamp of an encounter is not exposed by the Exposure Notification Framework (ENF) for privacy reasons afaik. Therefore CWA cannot show an exact time but only the day when the exposure occurred (see [here](https://github.com/corona-warn-app/cwa-backlog/issues/23#issuecomment-678692630) for more details on the response CWA gets from ENF). If you have a rooted android device you can bypass this restriction e.g. with the [corona-warn-companion-android](https://github.com/mh-/corona-warn-companion-android) though.

**Comment by Ein-Tim** [comment]
Could this Issue please be moved to the Documentation Repo since it affects both Android and iOS? 
(@abro1i @svengabr)

**Comment by vschoech** [comment]
> It's quite bad to not explicitly stating the time frame when the contacts occurred, since this way you could narrow it down how the contact occurred.

@RubenKelevra *It's not a bug, it's a feature.* The precise time of the encounter is intentionally not given away because you don't want to expose the person that has reported being infected. Anonymity is an important aspect for broad acceptance. If it is easy to trace back who reported an infection, less infections will be reported.

That said, intentionally not giving a precise time frame is not the same as giving a time frame that *seems* so be precise but is not consistent with other information (which is what this ticket is about).

**Comment by RubenKelevra** [comment]
I know that there are limits in time resolution, but the resolution which is possible technically should be shown.

Say if the contact is known to be occurred on Thursday, show Thursday - not 4 days ago. And if it's known to be in an 8 hour timeframe on Thursday, show 'between 3am and 11am'.

The current plan display is just unnecessarily confusing IMHO.

**Comment by svengabr** [comment]
Hello everyone,

there is already an internal proposal for showing more information regarding risk encounters. We will open a new wishlist entry for this soon to make it more transparent for everyone.

An overhaul of risk encounters information UI/UX was implemented with the app version 1.3.1 which is already available.
